Snmp observ lib

v-zhuravlev opened the pull request:
This is universal SNMP dashboard and alerts supporting multiple network vendors.

There are two tiers of dashboards (using layout you can find for Linux or Windows)
- `Fleet` to add network overview
- `Overview` to give more details about
- `logs` dashboard based on syslog received messages from devices.
 
This observ lib relies on snmp_exporter modules of version 0.28.0 or above. Please see README.md for information which snmp_exporter modules should be used to poll metric groups below:



Currently supported metrics groups:
- Interfaces stats like unicast,broadcast,multicasts,errors and octets in/out (all devices)
- CPU (mikrotik,dell,cisco,juniper,dlink,eltex,ubuquiti airos(old))
- Memory (mikrotik,dell,cisco,juniper,dlink,eltex,ubuquiti airos(old))
- sensors (temperature etc) (for Cisco and Juniper only for now)
- FRU status (for cisco , optional row)
- Fibre channel (for cisco MDS only, optional row)

Note: more vendors will added once they are supported by snmp_exporter. As well as more metric groups will be added to existing vendors.(like adding support for temperature sensors etc).

Alerts:
- Interfaces, Fibre channel errors, SNMP exporter, reboot, cpu and memory usage are controlled.

Annotations:
- Dashboards support annotations when relevant alerts start firing as well as critical alerts are found recevied from syslog.

Datalinks:
- Fleet table and fleet panels supports data links to drill down to instance of intereset.


Examples:
Fleet overview:
0) ![image](https://github.com/user-attachments/assets/ac0cb700-2386-4dfa-9212-c81d6175363a)
note: transmitted/received in table is empty is due to static counters from emulated devices. They actually work. :)

1) Cisco c2950:
![image](https://github.com/user-attachments/assets/6a2e2c6a-8bf0-4f94-b240-f018f270ce58)

2) Cisco 7600:
![image](https://github.com/user-attachments/assets/d5140dc0-96a9-496a-9ccb-c3ac70026acb)


3) Mikrotik
![image](https://github.com/user-attachments/assets/40ecb72d-53b4-44f1-bd6f-f733c62d58b6)
4) Mikrotik 2
![image](https://github.com/user-attachments/assets/3cc6ed68-78bd-4194-a719-7c02d193c2d7)

Dasomeone commented on snmp-observ-lib/signals/memory.libsonnet:
```diff
@@ -0,0 +1,142 @@
+          },
+          juniper: {
+            expr: 'jnxOperatingBuffer{jnxOperatingContentsIndex="9", %(queriesSelector)s}',
+            // aggKeepLabels: ['jnxOperatingDescr'],
```
Minor nit, if we don't want these aggregation labels they shouldn't be commented without a reason given

v-zhuravlev replied:
added comments. I want to use it in next iteration.

Dasomeone commented on snmp-observ-lib/panels/fleet.libsonnet:
```diff
@@ -0,0 +1,129 @@
+      + commonlib.panels.network.timeSeries.dropped.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsUnicast:
+      signals.fleetInterface.networkInUnicastPacketsPerSec.withTopK().asTimeSeries('Unicast')
+      + signals.fleetInterface.networkOutUnicastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.unicast.stylize()
+      + commonlib.panels.network.timeSeries.unicast.withDataLink(this.config.instanceLabels, this.grafana.dashboards['snmp-overview.json'].uid)
+      + commonlib.panels.network.timeSeries.unicast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsBroadcast:
+      signals.fleetInterface.networkInBroadcastPacketsPerSec.withTopK().asTimeSeries('Broadcast')
+      + signals.fleetInterface.networkOutBroadcastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.broadcast.stylize()
+      + commonlib.panels.network.timeSeries.broadcast.withDataLink(this.config.instanceLabels, this.grafana.dashboards['snmp-overview.json'].uid)
+      + commonlib.panels.network.timeSeries.broadcast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsMulticast:
+      signals.fleetInterface.networkInMulticastPacketsPerSec.withTopK().asTimeSeries('Multicast')
+      + signals.fleetInterface.networkOutMulticastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.multicast.stylize()
+      + commonlib.panels.network.timeSeries.multicast.withDataLink(this.config.instanceLabels, this.grafana.dashboards['snmp-overview.json'].uid)
+      + commonlib.panels.network.timeSeries.multicast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
```
Based on the screenshot I think these three panels would benefit from a table based legend. Thoughts?

v-zhuravlev replied:
it would make panels narrower if table is moved to the right, not so great IMHO. 
And since there are no calculations in the legend, it wouldn't make a lot of sense if changing the legend type to table and keeping at the bottom.
I would keep it as is for now.

Dasomeone replied:
That's fair, something to keep an eye as we see more busy deployments then!

Dasomeone commented on snmp-observ-lib/panels/fleet.libsonnet:
```diff
@@ -0,0 +1,129 @@
+
+{
+  new(signals, this):: {
+    deviceTable:
+      local fieldOverride = g.panel.table.fieldOverride;
+      signals.fleetInterface.interfacesCount.asTable(name='Network devices', format='table', filterable=true)
+      + g.panel.table.panelOptions.withDescription('Network devices overview.')
+      + signals.system.sysName.asTableColumn(format='table')
+      + signals.system.version.asTableColumn(format='table')
+      + signals.system.uptime.asTableColumn(format='table')
+      + commonlib.panels.system.table.uptime.stylizeByName('Uptime')
+      + signals.cpu.cpuUsage.asTableColumn(format='table')
+      + commonlib.panels.generic.table.percentage.stylizeByName('CPU utilization')
+      + signals.memory.memoryUsage.asTableColumn(format='table')
+      + commonlib.panels.generic.table.percentage.stylizeByName('Memory utilization')
+      + signals.fleetInterface.networkOutBitPerSec.asTableColumn(format='table')
+      + commonlib.panels.generic.table.coldHotGauge.stylizeByName('Transmitted')
+      + signals.fleetInterface.networkInBitPerSec.asTableColumn(format='table')
+      + commonlib.panels.generic.table.coldHotGauge.stylizeByName('Received')
+      + signals.fleetInterface.networkInErrorsPerSec.asTableColumn(format='table')
+      + signals.fleetInterface.networkOutErrorsPerSec.asTableColumn(format='table')
+      + signals.fleetInterface.networkInDroppedPerSec.asTableColumn(format='table')
+      + signals.fleetInterface.networkOutDroppedPerSec.asTableColumn(format='table')
+      + signals.fleetInterface.ifInUnknownProtos.asTableColumn(format='table')
+      + signals.alerts.alertsWarning
+        .withExprWrappersMixin(['count by (%s)(' % std.join(',', this.config.instanceLabels + this.config.groupLabels), ')'])
+        .asTableColumn(format='table')
+      + signals.alerts.alertsCritical
+        .withExprWrappersMixin(['count by (%s)(' % std.join(',', this.config.instanceLabels + this.config.groupLabels), ')'])
+        .asTableColumn(format='table')
+      + {
```
This is a very wide table (wider than the dashboard, in the screenshot, as it cuts off the "dropped in" column)

I noticed the job column is all the same, maybe we want to do a rewrite or collapse by default?

v-zhuravlev replied:
let me try something..

Dasomeone replied:
I noticed the change to use `networkAllErrorsPerSec`, could you show a screenshot of what that looks like?

v-zhuravlev replied:
yepp, see below

Dasomeone replied:
Awesome thank you! That looks much much better

Dasomeone commented on snmp-observ-lib/panels/interface.libsonnet:
```diff
@@ -0,0 +1,78 @@
+      + commonlib.panels.network.timeSeries.dropped.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsUnicast:
+      signals.interface.networkInUnicastPacketsPerSec.withTopK().asTimeSeries('Unicast')
+      + signals.interface.networkOutUnicastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.unicast.stylize()
+      + commonlib.panels.network.timeSeries.unicast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsBroadcast:
+      signals.interface.networkInBroadcastPacketsPerSec.withTopK().asTimeSeries('Broadcast')
+      + signals.interface.networkOutBroadcastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.broadcast.stylize()
+      + commonlib.panels.network.timeSeries.broadcast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
+
+    packetsMulticast:
+      signals.interface.networkInMulticastPacketsPerSec.withTopK().asTimeSeries('Multicast')
+      + signals.interface.networkOutMulticastPacketsPerSec.withTopK().asPanelMixin()
+      + commonlib.panels.network.timeSeries.multicast.stylize()
+      + commonlib.panels.network.timeSeries.multicast.withNegateOutPackets()
+      + g.panel.timeSeries.queryOptions.withInterval(this.config.minInterval),
```
Same comment as with Fleet, might benefit from a table legend below the panel

v-zhuravlev replied:
see above

Dasomeone reviewed (commented):
Couple comments to address, but overall this looks VERY nice! Fantastic job Vitaly!

v-zhuravlev commented on snmp-observ-lib/alerts.libsonnet:
```diff
@@ -0,0 +1,383 @@
+              keep_firing_for: '3m',
+            },
+            {
+              alert: 'SNMPInterfaceFCerrors',
+              expr: |||
+                      (%s) > 0
+                    |||
+                    % [
+                      this.signals.fiber.fcIfInvalidTxWords.asRuleExpression(),
+                    ],
+              labels: {
+                severity: 'warning',
+              },
```
thanks!

v-zhuravlev commented on snmp-observ-lib/README.md:
````diff
@@ -0,0 +1,42 @@
+jb install https://github.com/grafana/jsonnet-libs/snmp-observ-lib
+```
+
+## Example
````
updated

v-zhuravlev commented on snmp-observ-lib/README.md:
```diff
@@ -0,0 +1,42 @@
+
+
+## Links
+https://cric.grenoble.cnrs.fr/Administrateurs/Outils/MIBS/?oid=1.3.6.1.2.1.2.2.1.3
```
removed

v-zhuravlev commented on snmp-observ-lib/README.md:
````diff
@@ -0,0 +1,42 @@
+|tplink | TP-LINK | TPLINK-SYSINFO-MIB,HOST-RESOURCES-MIB,SNMPv2-MIB,TPLINK-SYSMONITOR-MIB,IF-MIB | T2600G-28TS | https://www.tp-link.com/en/support/download/t2600g-28ts/#MIBs_Files https://www.tp-link.com/ru/support/faq/1330/ |system,if_mib|
+|ubiquiti_airos | Ubiquiti AirOS | FROGFOOT-RESOURCES-MIB,HOST-RESOURCES-MIB,SNMPv2-MIB,IEEE802dot11-MIB,IF-MIB | NanoStation M5, UAP-LR |  |system,if_mib,ubiquiti_airos|
+
+## Import
+
+```sh
+jb init
+jb install https://github.com/grafana/jsonnet-libs/snmp-observ-lib
+```
````
good idea!

Dasomeone commented:
Thanks for addressing the comments @v-zhuravlev, all looks mostly good though I think two comments got hidden by Github (wish that could be disabled...) so if you could address those I'd be happy to approve :)

v-zhuravlev commented:
> Thanks for addressing the comments @v-zhuravlev, all looks mostly good though I think two comments got hidden by Github (wish that could be disabled...) so if you could address those I'd be happy to approve :)

1) New fleet table, now more compact with 5 columns merged:
<img width="2049" alt="image" src="https://github.com/user-attachments/assets/7c2476ba-d718-4ecd-87de-67ba2de1e7ff" />
Kept job though, as it is not always the case that grouplabels would be identical.
2) Updated interfaces table, dropped group/instance labels from there as table is intended to be used with single instance(device) only.
<img width="2036" alt="image" src="https://github.com/user-attachments/assets/2fc5d487-94e9-4c33-a13c-c13dd40e35bb" />

v-zhuravlev commented on snmp-observ-lib/config.libsonnet:
```diff
@@ -0,0 +1,54 @@
+
+  // logs lib related
+  enableLokiLogs: true,
+  logsFilteringSelector: 'job="syslog"',
```
that would be a filteringSelector, instancelabels and grouplabels would be added as well.

Right now, logs and metrics from network devices are out of sync, thats why we need separate set of selectors here. 

With other SUOs it is much easier to keep those labels in sync, as both signals can be collected with same alloy agent.

job=syslog is the default here to catch logs that come from syslog if other selectors are set to wildcards {job=~".*", instance=~".+"} (fleet dashboard default). Till we find better default to catch network logs only.

Dasomeone replied:
Gotcha, thanks for clarifying!

v-zhuravlev commented on snmp-observ-lib/alerts.libsonnet:
```diff
@@ -0,0 +1,383 @@
+                ||| % instanceLabel,
+              },
+              'for': '10m',
+              keep_firing_for: '3m',
```
updated to 5m

Dasomeone reviewed (approved):
Awesome work Vitaly, thank you for addressing all the feedback! LGTM :)
